From: Cyrus Lien <cyrus.lien@canonical.com>
Date: Wed, 1 Jul 2026 03:08:48 +0800
Subject: system,
 users: Keep fingerprint add-print popover visible when scaled

When configuring Fingerprint Login on small or tablet displays,
especially at 200% display scaling, clicking "Scan new fingerprint" can
produce no visible response. The finger-list popover is too tall to fit
below the button, GTK cannot place it above either, and the popover
fails to appear, blocking the user from selecting a finger to enroll.

Track the add-print button and choose the popover position by available
room. Wrap the popover content in a scrolled window so the popup can stay
visible when vertical space is limited.

Bug-Ubuntu: https://launchpad.net/bugs/2158603
Forwarded: not-needed

Upstream is working on redesign fingerprint management interface:
https://gitlab.gnome.org/GNOME/gnome-control-center/-/merge_requests/3399
---
 panels/system/users/cc-fingerprint-dialog.blp |  12 +-
 panels/system/users/cc-fingerprint-dialog.c   | 166 +++++++++++++++++++++++++-
 2 files changed, 174 insertions(+), 4 deletions(-)

diff --git a/panels/system/users/cc-fingerprint-dialog.blp b/panels/system/users/cc-fingerprint-dialog.blp
index ffb7b9b..4744283 100644
--- a/panels/system/users/cc-fingerprint-dialog.blp
+++ b/panels/system/users/cc-fingerprint-dialog.blp
@@ -250,9 +250,15 @@ template $CcFingerprintDialog: Adw.Dialog {
 Popover add_print_popover {
   position: bottom;
 
-  Box add_print_popover_box {
+  ScrolledWindow add_print_popover_scroller {
     can-focus: false;
-    spacing: 6;
-    orientation: vertical;
+    hscrollbar-policy: never;
+    propagate-natural-height: true;
+
+    Box add_print_popover_box {
+      can-focus: false;
+      spacing: 6;
+      orientation: vertical;
+    }
   }
 }
diff --git a/panels/system/users/cc-fingerprint-dialog.c b/panels/system/users/cc-fingerprint-dialog.c
index 71f804a..65561e1 100644
--- a/panels/system/users/cc-fingerprint-dialog.c
+++ b/panels/system/users/cc-fingerprint-dialog.c
@@ -62,6 +62,7 @@ struct _CcFingerprintDialog
   GtkButton      *delete_prints_button;
   GtkButton      *done_button;
   GtkBox         *add_print_popover_box;
+  GtkWidget      *add_print_popover_scroller;
   GtkEntry       *enroll_print_entry;
   GtkFlowBox     *prints_gallery;
   AdwHeaderBar   *titlebar;
@@ -74,6 +75,7 @@ struct _CcFingerprintDialog
   AdwSpinner     *spinner;
   GtkStack       *stack;
   GtkWidget      *add_print_icon;
+  GtkMenuButton  *add_print_button;
   GtkWidget      *device_selector;
   GtkWidget      *enroll_print_bin;
   GtkWidget      *enroll_result_icon;
@@ -149,6 +151,160 @@ const char * ENROLL_STATE_CLASSES[N_ENROLL_STATES] = {
 
 static GParamSpec *properties[N_PROPS];
 
+static GtkArrowType
+arrow_type_for_position (GtkPositionType position)
+{
+  switch (position)
+    {
+    case GTK_POS_LEFT:
+      return GTK_ARROW_LEFT;
+    case GTK_POS_RIGHT:
+      return GTK_ARROW_RIGHT;
+    case GTK_POS_TOP:
+      return GTK_ARROW_UP;
+    case GTK_POS_BOTTOM:
+    default:
+      return GTK_ARROW_DOWN;
+    }
+}
+
+static void
+set_add_print_popover_position (CcFingerprintDialog *self,
+                                GtkPositionType      position)
+{
+  gtk_menu_button_set_direction (self->add_print_button, arrow_type_for_position (position));
+  gtk_popover_set_position (self->add_print_popover, position);
+}
+
+static void
+set_add_print_popover_max_height (CcFingerprintDialog *self,
+                                  GtkPositionType      position,
+                                  int                  top_space,
+                                  int                  bottom_space)
+{
+  int max_height;
+
+  if (!self->add_print_popover_scroller)
+    return;
+
+  if (position == GTK_POS_TOP)
+    max_height = top_space;
+  else if (position == GTK_POS_BOTTOM)
+    max_height = bottom_space;
+  else
+    max_height = gtk_widget_get_height (GTK_WIDGET (self));
+
+  gtk_scrolled_window_set_max_content_height (GTK_SCROLLED_WINDOW (self->add_print_popover_scroller),
+                                              MAX (max_height - 12, 1));
+}
+
+static gboolean
+get_add_print_button_space (CcFingerprintDialog *self,
+                            int                 *left_space,
+                            int                 *right_space,
+                            int                 *top_space,
+                            int                 *bottom_space)
+{
+  GtkWidget *button = GTK_WIDGET (self->add_print_button);
+  graphene_rect_t bounds;
+  int root_height;
+  int root_width;
+
+  if (!gtk_widget_compute_bounds (button, GTK_WIDGET (self), &bounds))
+    return FALSE;
+
+  root_height = gtk_widget_get_height (GTK_WIDGET (self));
+  root_width = gtk_widget_get_width (GTK_WIDGET (self));
+  *left_space = MAX ((int) bounds.origin.x, 0);
+  *right_space = MAX (root_width - (int) (bounds.origin.x + bounds.size.width), 0);
+  *top_space = MAX ((int) bounds.origin.y, 0);
+  *bottom_space = MAX (root_height - (int) (bounds.origin.y + bounds.size.height), 0);
+
+  return TRUE;
+}
+
+static int
+get_add_print_popover_natural_width (CcFingerprintDialog *self)
+{
+  int minimum_width;
+  int natural_width;
+
+  gtk_widget_measure (GTK_WIDGET (self->add_print_popover_box),
+                      GTK_ORIENTATION_HORIZONTAL,
+                      -1,
+                      &minimum_width,
+                      &natural_width,
+                      NULL,
+                      NULL);
+
+  return MAX (minimum_width, natural_width);
+}
+
+static int
+get_add_print_popover_natural_height (CcFingerprintDialog *self)
+{
+  int minimum_height;
+  int natural_height;
+
+  gtk_widget_measure (GTK_WIDGET (self->add_print_popover_box),
+                      GTK_ORIENTATION_VERTICAL,
+                      -1,
+                      &minimum_height,
+                      &natural_height,
+                      NULL,
+                      NULL);
+
+  return MAX (minimum_height, natural_height);
+}
+
+static GtkPositionType
+get_add_print_popover_position (CcFingerprintDialog *self)
+{
+  int bottom_space;
+  int left_space;
+  int natural_height;
+  int natural_width;
+  int right_space;
+  int top_space;
+
+  if (!self->add_print_button ||
+      !get_add_print_button_space (self, &left_space, &right_space, &top_space, &bottom_space))
+    return GTK_POS_RIGHT;
+
+  natural_width = get_add_print_popover_natural_width (self);
+  natural_height = get_add_print_popover_natural_height (self);
+
+  if (natural_width <= 0 || right_space >= natural_width)
+    {
+      set_add_print_popover_max_height (self, GTK_POS_RIGHT, top_space, bottom_space);
+      return GTK_POS_RIGHT;
+    }
+
+  if (left_space >= natural_width)
+    {
+      set_add_print_popover_max_height (self, GTK_POS_LEFT, top_space, bottom_space);
+      return GTK_POS_LEFT;
+    }
+
+  if (natural_height <= 0 || bottom_space >= natural_height)
+    {
+      set_add_print_popover_max_height (self, GTK_POS_BOTTOM, top_space, bottom_space);
+      return GTK_POS_BOTTOM;
+    }
+
+  set_add_print_popover_max_height (self, GTK_POS_TOP, top_space, bottom_space);
+  return GTK_POS_TOP;
+}
+
+static void
+on_add_print_button_active_changed_cb (CcFingerprintDialog *self)
+{
+  if (!gtk_menu_button_get_active (self->add_print_button))
+    return;
+
+  set_add_print_popover_position (self, get_add_print_popover_position (self));
+}
+
 CcFingerprintDialog *
 cc_fingerprint_dialog_new (CcFingerprintManager *manager)
 {
@@ -1034,8 +1190,15 @@ populate_prints_gallery (CcFingerprintDialog *self)
 
   populate_add_print_popover (self);
   button = g_object_get_data (G_OBJECT (self->add_print_icon), "button");
-  gtk_menu_button_set_popover (GTK_MENU_BUTTON (button),
+  self->add_print_button = GTK_MENU_BUTTON (button);
+  gtk_menu_button_set_popover (self->add_print_button,
                                GTK_WIDGET (self->add_print_popover));
+  set_add_print_popover_position (self, GTK_POS_RIGHT);
+  g_signal_connect_object (self->add_print_button,
+                           "notify::active",
+                           G_CALLBACK (on_add_print_button_active_changed_cb),
+                           self,
+                           G_CONNECT_SWAPPED);
 
   gtk_flow_box_insert (self->prints_gallery, self->add_print_icon, -1);
   gtk_flow_box_set_max_children_per_line (self->prints_gallery, 1);
@@ -1494,6 +1657,7 @@ cc_fingerprint_dialog_class_init (CcFingerprintDialogClass *klass)
 
   gtk_widget_class_bind_template_child (widget_class, CcFingerprintDialog, add_print_popover);
   gtk_widget_class_bind_template_child (widget_class, CcFingerprintDialog, add_print_popover_box);
+  gtk_widget_class_bind_template_child (widget_class, CcFingerprintDialog, add_print_popover_scroller);
   gtk_widget_class_bind_template_child (widget_class, CcFingerprintDialog, back_button);
   gtk_widget_class_bind_template_child (widget_class, CcFingerprintDialog, cancel_button);
   gtk_widget_class_bind_template_child (widget_class, CcFingerprintDialog, delete_prints_button);
